internal/command/user_v2_human.go

```diff
@@ -0,0 +1,540 @@
+package command
+
+import (
+	"context"
+
+	"golang.org/x/text/language"
+
+	"github.com/zitadel/zitadel/internal/crypto"
+	"github.com/zitadel/zitadel/internal/domain"
+	"github.com/zitadel/zitadel/internal/errors"
+	"github.com/zitadel/zitadel/internal/eventstore"
+	"github.com/zitadel/zitadel/internal/repository/user"
+	"github.com/zitadel/zitadel/internal/telemetry/tracing"
+)
+
+type ChangeHuman struct {
+	ID       string
+	Username *string
+	Profile  *Profile
+	Email    *Email
+	Phone    *Phone
+
+	Password *Password
+
+	// Details are set after a successful execution of the command
+	Details *domain.ObjectDetails
+
+	// EmailCode is set by the command
+	EmailCode *string
+
+	// PhoneCode is set by the command
+	PhoneCode *string
+}
+
+type Profile struct {
+	FirstName         *string
+	LastName          *string
+	NickName          *string
+	DisplayName       *string
+	PreferredLanguage *language.Tag
+	Gender            *domain.Gender
+}
+
+type Password struct {
+	// Either you have to have permission, a password code or the old password to change
+	PasswordCode        *string
+	OldPassword         *string
+	Password            *string
+	EncodedPasswordHash *string
+
+	ChangeRequired bool
+}
+
+func (h *ChangeHuman) Validate(hasher *crypto.PasswordHasher) (err error) {
+	if h.Email != nil && h.Email.Address != "" {
+		if err := h.Email.Validate(); err != nil {
+			return err
+		}
+	}
+
+	if h.Phone != nil && h.Phone.Number != "" {
+		if h.Phone.Number, err = h.Phone.Number.Normalize(); err != nil {
+			return err
+		}
+	}
+
+	if h.Password != nil {
+		if err := h.Password.Validate(hasher); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p *Password) Validate(hasher *crypto.PasswordHasher) error {
+	if p.EncodedPasswordHash != nil {
+		if !hasher.EncodingSupported(*p.EncodedPasswordHash) {
+			return errors.ThrowInvalidArgument(nil, "USER-JDk4t", "Errors.User.Password.NotSupported")
+		}
+	}
+	if p.Password == nil && p.EncodedPasswordHash == nil {
+		return errors.ThrowInvalidArgument(nil, "COMMAND-3M0fs", "Errors.User.Password.Empty")
+	}
+	if p.Password != nil && p.EncodedPasswordHash != nil {
+		return errors.ThrowInvalidArgument(nil, "COMMAND-3M0fsss", "Errors.User.Password.NotSupported")
```

Comment:
I don't fully understand what's happening here. If both a password and an encoding password is provided, we return `NotSupported`. Shouldn't it be something like an "Invalid Argument"?

Reply:
You're right, the thrown error type ist wrong.

Reply:
Removed the error, as I think it doesn't matter how you provide the password and lastly the encoded password gets used if both are provided.